lib/node_modules/@stdlib/utils/parse-dsv/lib/parse_dsv.js

```diff
@@ -0,0 +1,196 @@
+const fs = require('fs');
+const Buffer = require('buffer').Buffer;
+
+const DQUOTE = '"';
+const CR = '\r';
+const LF = '\n';
+const CRLF = '\r\n';
+
+const token_enum = {
+    'DELIM': 0,
+    'DQUOTE': 1,
+    'CR': 2,
+    'LF': 3,
+    'CRLF': 4,
+    'TEXTDATA': 5
+}
+
+const state_enum = {
+    'INITIAL': 0,
+    'REST': 1,
+    'RECORD': 2,
+    'RECORDREST': 3,
+    'FIELD': 4,
+    'ESCAPED': 5,
+    'NONESCAPED': 6,
+    'TWODQUOTE': 7
+}
+
+function parse_dsv_stream(DELIM = ',') {
+
+    // assume delimiter is always 1 byte long
```

Comment:
How would accounting for multi-byte delimiters affect the implementation? Or stated otherwise, is it straightforward to generalize to delimiters spanning multiple bytes?

Reply:
I think that is leftover from the old implementation because I was transforming it to it's dec value with `charCodeAt(0)` to compare against. I imagine right now that the only problem is a potential hit to efficiency.

Whenever an array of characters is pushed via `incrparse(chars)`, `readNextTokens()` will always be called. Right now each token is 1 character (or 2 in the case of CRLF) long. This means that as long as 1 character that is not CR is pushed, a token can be generated and the `state` variable can be updated.

In the case of CR and CRLF, the implementation reads CR off the buffer and checks if there is more to be read, if there is at least 1 more character, it's possible to determine if the token should be CR or CRLF. However, if there isn't i.e. there was only 1 character in the buffer, then CR is placed back into the buffer and no token is appended. Needing 2 characters in the buffer for just this one case is ok in my opinion.

However, a delimiter of variable length might cause a problem. The buffer could contain up to `DELIM.length` characters before the next token could be determined and before that every character would be read multiple times.

I think it's technically possible to have a similar state variable for the tokenizer specifically to let it know that it was attempting to create a delimiter token and another buffer for what has already been processed to fix that problem. But if I did that for the delimiter I'd also want to do it for CR for consistency although it is not necessary.

Another possible solution is to just wait until the buffer is `DELIM.length` long before doing any kind of processing.

On another note, this reminds me that I'm not actually handling the unicode character in the case correctly because it's more than 1 byte and I use `charCodeAt(0)` to check if it falls under valid text data and I should be checking each byte. And I need to update the byte ranges because I know they're specifically for CSV which is why the value for the comma is left out.

This also brings up the discussion of what restrictions to place on a custom delimiter. I think having a delimiter that starts with any of the other specific cases in the switch in `readNextTokens()` would break it i.e. CR, LF, and DQUOTE and any other case that might be added like single quotes.

Reply:
@rei2hu I think the [Pandas](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html) API is informative here, both in its options and behavior. Notably, just as we might have TSV (tab-separated), stands to reason that we might encounter multi-space delimited values (e.g., 4 spaces to separate columns). The underlying Pandas implementation may also prove informative regarding what limits, if any, should be imposed on the delimiter.

Reply:
For (2), can append a method to `incrparse` called, say, `.end( [chunk] )` which can indicate that no more data is to be provided. And further attempts to write would result in an `Error`.

Alternatively, we can address (2) by allowing users to provide an indicator value to `incrparse`. Similar to Node.js streams, we can make that value `null`. Hence, if `incrparse( null )`, then parsing is completed and we throw an error if any further data is provided.

Reply:
I think (3) can be partially remedied by minimizing memory allocation (e.g., `splice()` ops). Iterating over an array (or array view) is fast and should not be a bottleneck. If we need to, we could branch based on delimiter length, where we have a more optimized version for single byte delimiters and a more general version for multi-byte delimiters.

Reply:
Regarding (1), does RC 4180 have any opinions on this?

Reply:
Re: 1

RFC 4180 is specifically for csvs so it specifies the delimiter should be a comma. I'm not able to find anything about dsvs. I did make a change to allow the delimiter to start with any character by moving the delim case out into an if statement. I do see problems by allowing special characters to be in the delimitter for example a delimiter of `""` might make things very weird.

Re: 2

I added a case for `null` although it will have to be sent in an array just to stick with consistency like `incrspace([null])` or `incrspace(['a',',','b',null])`

Re: 3

Good suggestion. I think I was fixated on the case where it's actually a delimiter and that's why I spliced early.

Reply:
Re: 1. Agreed. If we need to apply some restrictions, so be it. I would be curious what Pandas and other commonly used tools do in this case.

Reply:
Pandas doesn't seem to limit anything. But I think pandas works differently for certain cases. For example:
```py
d = "\"abaa\",\"bbab\",\"cbac\"\n\"1ba1\",\"2ba2\",\"3ba3\""                                            
csv = pd.read_csv(StringIO(d), sep=',', engine='python')                                                
print(csv)
print()

# output makes sense
# "abaa","bbab","cbac"
# "1ba1","2ba2","3ba3"
#    abaa  bbab  cbac
# 0  1ba1  2ba2  3ba3
```
However, if I change the `,` to `ab` in the input and set the delimitter to `ab` i.e.
```py
d = "\"abaa\"ba\"bbab\"ba\"cbac\"\n\"1ba1\"ba\"2ba2\"ba\"3ba3\""                                        
csv = pd.read_csv(StringIO(d), sep='ba', engine='python')                                               
print(csv)

# output replaces delimiter with blank space
# "abaa"ba"bbab"ba"cbac"
# "1ba1"ba"2ba2"ba"3ba3"
#    "a  a"  "b  b"  "c  c"
# 0  "1  1"  "2  2"  "3  3"
```
In my opinion, this is not supposed to happen because the fields are quoted so the delimiter is allowed inside. From RFC 4180
>    6.  Fields containing line breaks (CRLF), double quotes, and commas
       should be enclosed in double-quotes.
The idea should be extended to the delimiter instead of commas.

I also did test `""` as a separator and it's allowed although NaN shows up for some reason. I might be missing something though.
```py
d = "\"\"\"\"\"\"a\n\"\"\"\"\"\"z"
csv = pd.read_csv(StringIO(d), sep='""', engine='python')
print(d)
print(csv)

# """"""a
# """"""z
#    Unnamed: 0  Unnamed: 1  Unnamed: 2  a
# 0         NaN         NaN         NaN  z
```

Reply:
@rei2hu Agreed. Pandas behavior there does not seem correct. And regarding the `""` separator case, yeah, seems like it is "allowed", but the result is gibberish. I think excluding `""` from being a separator is fine. If people are creating bizarro data formats (i.e., using `""` as the separator, which is, I think, clearly not in the spirit of CSV) and wanting it to be parsed perfectly as CSV, then it is on them.